### PR TITLE
ref(query): Add optional `format` parameter to `ClickhouseQuery.format_sql`

### PR DESCRIPTION
**Pull request by tkaemming**
This allows the `Reader[ClickhouseQuery]` to change the formatter used by ClickHouse by adding the 
`FORMAT` clause to the query. This is required for the HTTP reader, since it expects data in the `JSON` format.

This is kind of clunky, but seemed better than the alternatives that I considered:

1. Just tacking this onto the end of the string in the `Reader` when necessary. Seemed like a bad idea, since it splits the query formatting across two classes and relies on the assumption that the `FORMAT` clause is always the last one in the `SELECT` grammar.
2. Adding the format to the `Query` object. This would mean introducing an API for updating the format to `ClickhouseQuery` for this special case. Seemed like a bad idea, since it would set the precedent for the `Reader` to actually mutate the `ClickhouseQuery` object (and also is just a lot of work.)
3. Always adding `FORMAT JSON` to the formatted query, as in the original draft implementation of the HTTP reader (#406.) Seemed like a bad idea, since though the native driver ignores it as of today, there are no guarantees that it will ignore it forever. It would also require changing a lot of (unrelated, in my opinion) tests.

### CODE REVIEW

File: snuba/clickhouse/query.py [hunk 1]
```diff
@@ -19,7 +17,7 @@ class ClickhouseQuery(ABC):
     """
 
     @abstractmethod
-    def format_sql(self) -> str:
+    def format_sql(self, format: Optional[str] = None) -> str:
```
Comment (tkaemming) [inline snuba/clickhouse/query.py]: There's a bunch of stuff in this file that Black wants to reformat, I skipped checks to avoid noise in this patch.

**Review by fpacifici — COMMENTED** [review]
I assume the format will be picked by the reader, is that correct?
If yes I don't think this is a wrong or clunky direction. If we look at the main responsibility of ClickhouseQuery as managing the ClickhouseAST, doing query processing and rely on a visitor to format the query, later we could allow the Reader to instantiate the visitor and make the FORMAT (or other reader parameters) instantiation parameters of the visitor itself instead of hardcoding the instantiation at line 55 of astquery.py

What do you think about this direction? Of course it can be implemented only after: AST in prod, split snuba query/clickhosue query, extend the visitor to format the query and not only expressions.

Could you also please see my comment around the template method.

File: snuba/clickhouse/query.py [hunk 2]
```diff
@@ -145,6 +143,9 @@ def __init__(
...
         )
 
-    def format_sql(self) -> str:
+    def format_sql(self, format: Optional[str] = None) -> str:
         """Produces a SQL string from the parameters."""
-        return self.__formatted_query
+        query = self.__formatted_query
+        if format is not None:
+            query = f"{query} FORMAT {format}"
+        return query
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/clickhouse/query.py]: I'm fine with making this change but generally see this pattern as a bit of a smell (though I don't have a good rationale as to why, it just usually seems that it is usually a half-measure better served by a different abstraction.) Maybe this is justification enough for later implementing the visitor/formatter class that you described in the overall comment.

**Review by tkaemming — COMMENTED** [review]
> I assume the format will be picked by the reader, is that correct?

Yes, correct.

> If yes I don't think this is a wrong or clunky direction. If we look at the main responsibility of ClickhouseQuery as managing the ClickhouseAST, doing query processing and rely on a visitor to format the query, later we could allow the Reader to instantiate the visitor and make the FORMAT (or other reader parameters) instantiation parameters of the visitor itself instead of hardcoding the instantiation at line 55 of astquery.py

I think there might be a piece that is missing here or I am misunderstanding what you're suggesting here. The visitor isn't aware of the structure of the query itself, just how the component pieces of an expression relate to each other. The value of the `FORMAT` clause is not an expression, so I'm not sure how it would relate here.

> What do you think about this direction? Of course it can be implemented only after: AST in prod, split snuba query/clickhosue query, extend the visitor to format the query and not only expressions.

I think this may answer my question above somewhat. Do you anticipate other possible ways of formatting queries beyond their string representation? Or do you feel that the responsibilities (structure vs. formatting) are substantial enough to justify separating them? (I don't have an opinion here, just trying to understand the thought process a bit better.)

**Comment by fpacifici** [comment]
_"I think there might be a piece that is missing here or I am misunderstanding what you're suggesting here. The visitor isn't aware of the structure of the query itself,"_

Indeed it is not (now).  It requires this to be done as well:
"extend the visitor to format the query and not only expressions"

While we did not discuss this in depth, I always thought that the AST to represent expressions was only one step towards a good query representation and eventually the AST should be extended to represent the full query (I have no plan for attempting this anytime soon).  As in general it is possible to express the whole SQL query with a single grammar. That would allow a query processor and a formatter to have a better context on the data they are working onto. 
In this case we could easily extend the visitor.
Since we never achieve the grand design we have in mind, technically nothing prevent the visitor to visit the Query object instead of the expression, which I think could be a good idea so that the formatting would not be done in this brittle way half through ad hoc code in ClickhouseAstQuery and half through the visitor.
When the visitor goes through the full query then your problem is solved.

File: snuba/clickhouse/query.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/clickhouse/query.py]: Personal opinion: I think the template method can become a smell when it is used for classes that do not really belong to a hierarchy or where composition should be used instead. But if the classes fit reasonably in a hierarchy, and there is no obvious alternative abstraction, the value of avoiding code duplication is a good incentive.

File: snuba/clickhouse/query.py [hunk 3]
```diff
@@ -50,14 +60,14 @@ def __init__(
             column_expr(dataset, util.tuplify(colname), query, parsing_context)
             for colname in column_names
         ]
-        select_clause = u"SELECT {}".format(
```
Comment (fpacifici) [inline snuba/clickhouse/query.py]: it is quite weird this was still around.

**Review by fpacifici — APPROVED** [review]

File: snuba/clickhouse/query.py (hunk 3 above)
Comment (tkaemming) [reply snuba/clickhouse/query.py]: Oops, this was the change I was referring to here: https://github.com/getsentry/snuba/pull/790#discussion_r382352395 

I wasn't paying attention to the actual diff on the last update and didn't avoid staging it. Oh well.

**Pull request merged**